File: example/call_client.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int main(int argc, char **argv) {
    int p[2], r, i;
    char buf[64];
    pid_t pid;

    if (argc != 3) {
        fprintf(stderr, "%s ARG1 ARG2\n", argv[0]);
        exit(EXIT_FAILURE);
    }

    char *request[] = {
        "../bin/client", "-c", "concat", "-ret", "-str", "-str", argv[1],
        "-str", argv[2], NULL
    };

    printf("Request: \n  ");
    for (i = 0; i < 9; i++) {
        printf("%s ", request[i]);
    }
    printf("\n\nAnswer : \n");

    if (pipe(p) != 0) {
        perror("pipe()");
        exit(EXIT_FAILURE);
    }

    pid = fork();
    if (pid < 0) { /* error */
        perror("fork()");
        exit(EXIT_FAILURE);
    } else if (pid == 0) { /* son */
        close(p[0]);

        if (dup2(p[1], STDOUT_FILENO) == -1) {
            perror("dup2()");
            exit(EXIT_FAILURE);
        }

        if (execvp(request[0], request) == -1) {
            perror("execvp()");
            exit(EXIT_FAILURE);
        }
    }


    close(p[1]);
    while((r = read(p[0], buf, 64)) > 0) {
        write(STDOUT_FILENO, buf, r);
    }

    if (r == -1) {
        perror("read()");
        exit(EXIT_FAILURE);
    }

    wait(NULL);

    return EXIT_SUCCESS;
}

File: example/callClient.ml

open Unix

let read_and_show fd buf =
  let length = Bytes.length buf in
  let rec aux () =
    match read fd buf 0 length with
    | -1 -> failwith "read_and_show"
    | 0  -> ()
    | n  -> ignore (write stdout buf 0 n); aux ()
  in aux ()

let _ =
  match Array.length Sys.argv with
  | 3 ->
    ()
  | _ ->
    Printf.eprintf "%s ARG1 ARG2\n" Sys.argv.(0);
    exit 1

let _ =
  let request = [|
    "../bin/client"; "-c"; "concat"; "-ret"; "-str"; 
    "-str"; Sys.argv.(1); "-str"; Sys.argv.(2)
  |] in
  let infd, outfd = pipe () in
  match fork () with
  | -1 ->
    failwith "fork error"
  | 0  ->
    close infd;
    dup2 outfd stdout;
    execv request.(0) request
  | n  ->
    close outfd;
    let buffer = Bytes.create 64 in
    read_and_show infd buffer;
    ignore (Unix.wait ())
